src/ray/core_worker/task_manager.h

```diff
@@ -139,16 +200,16 @@ class TaskManager : public TaskFinisherInterface {
   /// Protects below fields.
   mutable absl::Mutex mu_;
 
-  /// Map from task ID to a pair of:
-  ///   {task spec, number of allowed retries left}
-  /// This map contains one entry per pending task that we submitted.
+  /// This map contains one entry per pending task that we submitted. If
+  /// lineage_pinning_enabled_ is true, then this map also contains entries for
+  /// tasks that we submitted that have completed execution, but that may be
+  /// retried again in the future.
   /// TODO(swang): The TaskSpec protobuf must be copied into the
   /// PushTaskRequest protobuf when sent to a worker so that we can retry it if
   /// the worker fails. We could avoid this by either not caching the full
   /// TaskSpec for tasks that cannot be retried (e.g., actor tasks), or by
   /// storing a shared_ptr to a PushTaskRequest protobuf for all tasks.
-  absl::flat_hash_map<TaskID, std::pair<TaskSpecification, int>> pending_tasks_
-      GUARDED_BY(mu_);
+  absl::flat_hash_map<TaskID, TaskEntry> pending_tasks_ GUARDED_BY(mu_);
```

Comment:
Consider `past_tasks_`

Reply:
Could also considering having two separate maps if it doesn't complicate the code

Reply:
Hmm I prefer keeping them in the same map because there are already cases in this PR where we'll then have to check both maps and there will be more once it's possible to resubmit tasks.

`past_tasks_` isn't quite accurate either though since the map includes pending tasks. Maybe just `submitted_tasks_`?